lib/libxdp/xsk.c

```diff
@@ -333,16 +342,19 @@ int xsk_umem__create_with_fd(struct xsk_umem **umem_ptr, int fd,
 
 	umem->umem_area = umem_area;
 	INIT_LIST_HEAD(&umem->ctx_list);
-	xsk_set_umem_config(&umem->config, usr_config);
+	xsk_set_umem_config(&umem->config, opts);
 
 	memset(&mr, 0, sizeof(mr));
 	mr.addr = (uintptr_t)umem_area;
 	mr.len = size;
 	mr.chunk_size = umem->config.frame_size;
 	mr.headroom = umem->config.frame_headroom;
 	mr.flags = umem->config.flags;
+	mr.tx_metadata_len = OPTS_GET(opts, tx_metadata_len, XSK_UMEM__DEFAULT_TX_METADATA_LEN);
+	err = setsockopt(umem->fd, SOL_XDP, XDP_UMEM_REG, &mr, 
+		mr.tx_metadata_len? sizeof(mr) :
+		offsetof(struct xdp_umem_reg, tx_metadata_len));
```

Comment:
Putting the size conditional into the last argument of setsockopt makes the code very dense and hampers readability. Let's pull it out into a variable, maybe defined as:
```c
size_t mr_size = sizeof(mr);
```

And then we can have a conditional with a comment, like:
```c
/* Older kernels don't support tx_metadata_len, skip if we are not setting a value */
if (!mr.tx_metadata_len)
    mr_size = offsetof(struct xdp_umem_reg, tx_metadata_len);
```